src/coreclr/vm/runtimehandles.cpp

```diff
@@ -1863,8 +1863,7 @@ extern "C" void QCALLTYPE RuntimeMethodHandle_StripMethodInstantiation(MethodDes
 //          async variants for task-returning methods
 //
 // For {task-returning, async} variants Reflection hands out only the task-returning variant.
-//  the async varinat is an implementation detail that conceptually does not exist.
-//  TODO: (async) the filtering may not cover all scenarios. Review and add tests.
+//  the async variant is an implementation detail that conceptually does not exist.
```

Comment:
Does `MethodBase.GetCurrentMethod()` work correctly inside async methods?

Reply:
> Does `MethodBase.GetCurrentMethod()` work correctly inside async methods?

Yes. I have added tests in this PR. For all kind of stack situations like inside async, Task-returning, one stack level up...

There is a test for `MethodBase.GetCurrentMethod()` as well. For both async and ordinary Task-returning (just in case).

Reply:
Yes - on CoreCLR. On NativeAOT it may work, in theory, but not right now, tracking issue - https://github.com/dotnet/runtime/issues/122546

Reply:
Actually there is a flip. It is in `RuntimeMethodHandle_GetStubIfNeededSlow`

https://github.com/dotnet/runtime/blob/20805161391d45095f4bcf27d0483f600e608b10/src/coreclr/vm/runtimehandles.cpp#L1930-L1934

The earlier fixes for reflection relied on that too. 

`GetStubIfNeededInternal` blocks async variants (intentionally, added in #118045). 
https://github.com/dotnet/runtime/blob/20805161391d45095f4bcf27d0483f600e608b10/src/coreclr/vm/runtimehandles.cpp#L1902-L1904

Then the resolution hits the slow path, which does the flip.
Then the result is cached by the caller.

Reply:
ok